JENKINS-57344: Whitelist requests for some java.io and other misc calls

haridsv opened the pull request:
This PR whitelists a few java.io reader and writer methods that are useful on filters (wrapper readers and writers). It also whitelists a few list operations

jglick commented on src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist:
```diff
@@ -34,13 +34,31 @@ method groovy.lang.Closure setResolveStrategy int
 
 # Useful general stuff:
 new java.io.StringReader java.lang.String
+method java.io.StringReader close
```
```
Signatures out of order, at index 34 found new java.io.StringReader java.lang.String but expected method java.io.StringReader close
Signatures out of order, at index 35 found method java.io.StringReader close but expected new java.io.StringReader java.lang.String
Signatures out of order, at index 47 found method java.io.StringWriter write char[] but expected method java.io.StringWriter int
Signatures out of order, at index 48 found method java.io.StringWriter write char[] int int but expected method java.io.StringWriter write char[]
Signatures out of order, at index 49 found method java.io.StringWriter int but expected method java.io.StringWriter write char[] int int
Signatures out of order, at index 293 found new java.util.LinkedHashSet but expected method java.util.Iterator hasNext
Signatures out of order, at index 294 found method java.util.Iterator hasNext but expected method java.util.Iterator next
Signatures out of order, at index 295 found method java.util.Iterator next but expected new java.util.LinkedHashSet
```

jglick commented on src/main/resources/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/generic-whitelist:
```diff
@@ -34,13 +34,31 @@ method groovy.lang.Closure setResolveStrategy int
 # Useful general stuff:
 new java.io.StringReader java.lang.String
+method java.io.StringReader close
+method java.io.StringReader read
```
You rather want to whitelist the _defining_ method, in `Reader`. Ditto in other cases.

haridsv replied:
Missed to respond to this earlier. That would open up those methods for all readers, right? Won't that whitelist too much, e.g., reading local files etc.?

haridsv replied:
Looks like specific overridden methods can't be whitelisted, so it is all or nothing.

haridsv replied:
I pushed a new change with the right attribution. I am assuming it is still OK because as long as the sandbox prevents the construction of those objects, there is not going to be a security issue with the broader whitelisting.

jglick replied:
> reading local files

You would first need to construct a `FileReader`, which would not be allowed.

bitwiseman commented:
@haridsv Do you plan to update this PR?

haridsv commented:
@bitwiseman Apologies for the delay, I will try to look into this tomorrow. I need to run this test locally to see the error and I think I did that a long time ago just need to find the steps again.

haridsv commented:
Some whitelist behavior seems to have changed. Previously, something like `"/a/b/c".split("/")[0..-2]` failed because of lack of whitelist for `getAt java.lang.Object[] groovy.lang.Range`, so I added that in PR. Now, the exact same Groovy snippet complains about the lack of whitelist for a different signature, which is `getAt java.lang.Object[] java.util.Collection`. It looks like the parameter type is `groovy.lang.IntRange` and it implements both `groovy.lang.Range<Integer>` and `java.util.Collection<Integer>`, so how is the interface chosen for creating the signature? Why did it switch from using `Range` to `Collection` ? Is this an indication of a bug in creating the signature in a deterministic way?

haridsv commented:
@dwnusbaum How to mark the jira as fixed? I probably don't have privileges to edit that field.
